File: iq/28/upper/find.go
// Copyright 2026 Samvel Khalatyan. All rights reserved.
//
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

package upper

func Find(nn []int, x int) (ub int, ok bool) {
	left := 0
	right := len(nn)
	for left < right {
		mid := left + (right-left)/2
		n := nn[mid]
		switch {
		case x < n:
			right = mid
		case x > n:
			left = mid + 1
		default:
			mid += 1
			if mid == len(nn) {
				return
			}
			return nn[mid], true
		}
	}
	if left == len(nn) {
		return
	}
	return nn[left], true
}

File: iq/28/upper/find.cc
// Copyright 2025 Samvel Khalatyan. All rights reserved.
//
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.
//
// clang-format off-next-line
//go:build ignore

#include "iq/28/upper/find.h"

#include <vector>

#include "absl/status/status.h"
#include "absl/status/statusor.h"

namespace iq::upper {

absl::StatusOr<int> Find([[maybe_unused]] const std::vector<int>& nn,
                         [[maybe_unused]] int k) {
  std::size_t low = 0;
  std::size_t hi = nn.size();
  while (low < hi) {
    const std::size_t idx = low + (hi - low) / 2;
    if (const int n = nn[idx]; k < n) {
      hi = idx;
    } else if (k > n) {
      low = idx + 1;
    } else {
      hi = idx + 1;
      break;
    }
  }
  if (hi == nn.size()) {
    return absl::OutOfRangeError("no such entry");
  }
  return nn[hi];
}

}  // namespace iq::upper
